Update logic for getting `Analytics` instance from package:unified_analytics

eliasyishak opened the pull request:
Part of:
- https://github.com/flutter/flutter/issues/128251

Currently, when we want to use the analytics instance from `package:unified_analytics`, we are just grabbing it from globals. However, with the legacy analytics instance, there are some things we check to return a no-op version of the instance.. for example, if we are running on bots or a non standard branch, we use a no-op instance

This PR uses the same previous checks for the new analytics instance

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

eliasyishak committed:
Move the logic into context runner

eliasyishak committed:
Use function instead of handling logic in context runner

eliasyishak commented on packages/flutter_tools/lib/src/reporting/unified_analytics.dart:
```diff
@@ -0,0 +1,36 @@
+
+import '../globals.dart' as globals;
+import '../version.dart';
+
```
The logic for this function is the same as to what was used for package:usage.

https://github.com/eliasyishak/flutter/blob/use-same-analytics-logic-as-pkg-usage/packages/flutter_tools/lib/src/reporting/usage.dart#L152-L163

eliasyishak committed:
DI parameters for `getAnalytics` function

christopherfujino commented on packages/flutter_tools/lib/src/context_runner.dart:
```diff
@@ -88,11 +89,12 @@ Future<T> runInContext<T>(
-        flutterChannel: globals.flutterVersion.channel,
-        flutterVersion: globals.flutterVersion.frameworkVersion,
-        dartVersion: globals.flutterVersion.dartSdkVersion,
+      Analytics: () => getAnalytics(
+        runningOnBot: runningOnBot,
+        flutterVersion: globals.flutterVersion,
+        suppressEnvFlag:
+            globals.platform.environment['FLUTTER_SUPPRESS_ANALYTICS'] ==
+                'true',
       ),
```
it occurs to me, this is already a function, you can just inline the logic here:

```suggestion
      Analytics: () {
        final String version = flutterVersion.getVersionString(redactUnknownBranches: true);
	if (
          // Ignore local user branches.
          version.startsWith('[user-branch]') ||
          // Many CI systems don't do a full git checkout.
          version.endsWith('/unknown') ||
          // Ignore bots.
          runningOnBot ||
          // Ignore when suppressed by FLUTTER_SUPPRESS_ANALYTICS.
          suppressEnvFlag,
        ) {
          return NoOpAnalytics();
        }
        return Analytics(
          tool: DashTool.flutterTool,
          flutterChannel: flutterVersion.channel,
          flutterVersion: flutterVersion.frameworkVersion,
          dartVersion: flutterVersion.dartSdkVersion,
        );
      },
```

christopherfujino replied:
also note you must follow the flutter style guide here

eliasyishak replied:
Lol actually if you look at the first commit on this PR, I had it like your suggestion but I used the function instead to match the other generators a bit more. But good to know I can have the logic in the file

christopherfujino commented on packages/flutter_tools/lib/src/context_runner.dart:
```diff
@@ -88,11 +89,12 @@ Future<T> runInContext<T>(
+        runningOnBot: runningOnBot,
+        flutterVersion: globals.flutterVersion,
+        suppressEnvFlag:
+            globals.platform.environment['FLUTTER_SUPPRESS_ANALYTICS'] ==
```
Please review the style guide on formatting https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting, dart format formatting is not allowed in flutter/flutter

christopherfujino replied:
also, I would recommend just passing `globals.platform.environment` to `getAnalytics()`, and then allowing that function to pull out the env variable it needs. also, should we call .toLowerCase() on the env variable before comparing with `'true'`?

eliasyishak replied:
> I would recommend just passing globals.platform.environment to getAnalytics()
That's a good idea! And agreed on making the env variable lowercase

christopherfujino commented on packages/flutter_tools/lib/src/reporting/unified_analytics.dart:
```diff
@@ -0,0 +1,36 @@
+  if (
+      // Ignore local user branches.
+      version.startsWith('[user-branch]') ||
+          // Many CI systems don't do a full git checkout.
+          version.endsWith('/unknown') ||
+          // Ignore bots.
+          runningOnBot ||
+          // Ignore when suppressed by FLUTTER_SUPPRESS_ANALYTICS.
+          suppressEnvFlag) {
```
please align these with the previous line

eliasyishak committed:
Pass env map to function + clean up from dartfmt

eliasyishak committed:
Include default parameter for `enableAsserts`

This will be useful when we want to validate that the events sent are in line with GA4's limitations which are checked within `package:unified_analytics`

eliasyishak committed:
Create unified_analytics_test.dart

eliasyishak committed:
`flutter update-packages` to get unified analytics 4.0.0

eliasyishak committed:
Revert "`flutter update-packages` to get unified analytics 4.0.0"

This reverts commit f2181ae73b2d6fa9c7c86c5fb1e7c794ca9aa839.

eliasyishak committed:
Running update pkgs after manually editing pubspec

Per the instructions in the wiki https://github.com/flutter/flutter/wiki/The-flutter-tool#adding-removing-or-making-changes-to-dart-dependencies

eliasyishak committed:
Revert "Running update pkgs after manually editing pubspec"

This reverts commit 2cb230056316006e2e209976a49c441e677f2911.

eliasyishak committed:
Merge remote-tracking branch 'upstream/master' into use-same-analytics-logic-as-pkg-usage

eliasyishak committed:
Update packages after merge with master

christopherfujino commented on dev/a11y_assessments/pubspec.yaml:
```diff
@@ -11,7 +11,7 @@ dependencies:
   collection: 1.18.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   material_color_utilities: 0.5.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  meta: 1.10.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  meta: 1.9.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```
drive by comment: downgrading meta looks bad...we should never downgrade dependencies

eliasyishak replied:
i think that may have been for one of the reverts i had done

christopherfujino replied:
Cool, please disregard

eliasyishak committed:
Added tests to test `getAnalytics` logic

eliasyishak committed:
Remove analyze error

eliasyishak committed:
Revert "Update packages after merge with master"

This reverts commit 86848dd352d28abc6a38d6fd4170988b6a9ba501.

eliasyishak committed:
Merge branch 'master' into use-same-analytics-logic-as-pkg-usage

christopherfujino commented:
FYI I'm unsubscribing for now so I don't get notified each time you push. Feel free to ping me again when it's ready for review.

eliasyishak committed:
Merge branch 'master' into use-same-analytics-logic-as-pkg-usage

eliasyishak committed:
`testUsingContext` -> `testWithoutContext`

eliasyishak committed:
For real instance, check it is not noop id

eliasyishak committed:
Static getter to check client id as well runtime type

eliasyishak commented:
@christopherfujino this PR is ready for your review now

eliasyishak committed:
Merge branch 'master' into use-same-analytics-logic-as-pkg-usage

christopherfujino commented on packages/flutter_tools/test/general.shard/unified_analytics_test.dart:
```diff
@@ -0,0 +1,128 @@
+        analyticsOverride: analyticsOverride,
+      );
+
+      expect(analytics.clientId, isNot(NoOpAnalytics.staticClientId),
+          reason: 'The CLIENT ID should be a randomly generated id');
```
And similar throughout this library

```suggestion
      expect(
        analytics.clientId,
        isNot(NoOpAnalytics.staticClientId),
        reason: 'The CLIENT ID should be a randomly generated id',
      );
```

christopherfujino commented on packages/flutter_tools/test/general.shard/unified_analytics_test.dart:
```diff
@@ -0,0 +1,128 @@
+
+      expect(analytics.clientId, isNot(NoOpAnalytics.staticClientId),
+          reason: 'The CLIENT ID should be a randomly generated id');
+      expect(analytics is! NoOpAnalytics, true);
```
When possible, prefer a specific matcher over expecting against a literal value, as the matchers will give more friendly failure messages (in this example, it will say what type your actual value is)

```suggestion
      expect(analytics, isNot(isA<NoOpAnalytics>()));
```

And throughout this library

christopherfujino commented on packages/flutter_tools/test/general.shard/unified_analytics_test.dart:
```diff
@@ -0,0 +1,128 @@
+      final Analytics analytics = getAnalytics(
+        runningOnBot: false,
+        flutterVersion: FakeFlutterVersion(),
+        environment: <String, String>{},
```
uber-nit: all these map literals can be const

```suggestion
        environment: const <String, String>{},
```

christopherfujino commented on packages/flutter_tools/test/general.shard/unified_analytics_test.dart:
```diff
@@ -0,0 +1,128 @@
+  });
+
+  group('Unit testing getAnalytics', () {
+    testWithoutContext('Successfully creates the instance for standard branch',
+        () {
```
```suggestion
    testWithoutContext('Successfully creates the instance for standard branch', () {
```

christopherfujino commented on packages/flutter_tools/lib/src/reporting/unified_analytics.dart:
```diff
@@ -0,0 +1,52 @@
+
+  // Providing an override of the [Analytics] instance is preferred when
+  // running tests for this function to prevent writing to the filesystem
+  if (analyticsOverride != null) {
```
This concerns me if this is the only safeguard that we have that we're not accessing the real filesystem and making real network requests. Let's discuss offline.

christopherfujino replied:
I think it will be sufficient if we ensure the `testUsingContext()` function overrides this value in the context.

eliasyishak committed:
Nit fixes

eliasyishak committed:
Update default context for `Analytics` in tests

eliasyishak commented on packages/flutter_tools/test/src/context.dart:
```diff
@@ -120,6 +121,7 @@ void testUsingContext(
           Pub: () => ThrowingPub(), // prevent accidentally using pub.
           CrashReporter: () => const NoopCrashReporter(),
           TemplateRenderer: () => const MustacheTemplateRenderer(),
+          Analytics: () => NoOpAnalytics(),
```
Added default here @christopherfujino

christopherfujino replied:
:+1:

christopherfujino reviewed (approved):
LGTM
